device_plugins/nvidia_gpu/nvidia_gpu.go

```diff
@@ -0,0 +1,281 @@
+	files, err := ioutil.ReadDir(devDirectory)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		if reg.MatchString(f.Name()) {
+			fmt.Printf("Found Nvidia GPU %q\n", f.Name())
+			ngm.devices[f.Name()] = pluginapi.Device{f.Name(), pluginapi.Healthy}
+		}
+	}
+
+	return nil
+}
+
+func (ngm *nvidiaGPUManager) GetDeviceState(DeviceName string) string {
+	// TODO: calling Nvidia tools to figure out actual device state
+	return pluginapi.Healthy
+}
+
+// Discovers Nvidia GPU devices, installs device drivers, and sets up device
+// access environment.
+func (ngm *nvidiaGPUManager) Start() error {
+	// Install Nvidia device drivers.
+	// TODO: turn this into a go library to get better error reporting.
+	fmt.Printf("Run installer script\n")
+	cmd := exec.Command("/bin/sh", "-c", "/usr/bin/nvidia-installer.sh")
+	stdoutStderr, err := cmd.CombinedOutput()
+	fmt.Printf("%s\n", stdoutStderr)
+
+	if _, err := os.Stat(nvidiaCtlDevice); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(nvidiaUVMDevice); err != nil {
+		return err
+	}
+	ngm.defaultDevices = []string{nvidiaCtlDevice, nvidiaUVMDevice}
+	_, err = os.Stat(nvidiaUVMToolsDevice)
+	if !os.IsNotExist(err) {
+		ngm.defaultDevices = append(ngm.defaultDevices, nvidiaUVMToolsDevice)
+	}
+
+	if err := ngm.discoverGPUs(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Register(kubeletEndpoint, pluginEndpoint, resourceName string) error {
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	defer conn.Close()
+	if err != nil {
+		return fmt.Errorf("device-plugin: cannot connect to kubelet service: %v", err)
+	}
+	client := pluginapi.NewRegistrationClient(conn)
+	reqt := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     pluginEndpoint,
+		ResourceName: resourceName,
+	}
+
+	_, err = client.Register(context.Background(), reqt)
+	if err != nil {
+		return fmt.Errorf("device-plugin: cannot register to kubelet service: %v", err)
+	}
+	return nil
+}
+
+// Implements DevicePlugin service functions
+func (ngm *nvidiaGPUManager) ListAndWatch(emtpy *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
+	fmt.Printf("device-plugin: ListAndWatch start\n")
+	changed := true
+	for {
+		for id, dev := range ngm.devices {
+			state := ngm.GetDeviceState(id)
+			if dev.Health != state {
+				changed = true
+				dev.Health = state
+				ngm.devices[id] = dev
+			}
+		}
+		if changed {
+			resp := new(pluginapi.ListAndWatchResponse)
+			for _, dev := range ngm.devices {
+				resp.Devices = append(resp.Devices, &pluginapi.Device{dev.ID, dev.Health})
+			}
+			fmt.Printf("ListAndWatch: send devices %v", resp)
+			if err := stream.Send(resp); err != nil {
+				fmt.Printf("device-plugin: cannot update device states: %v\n", err)
+				ngm.grpcServer.Stop()
```

Comment:
What happens if the server is stopped?

Reply:
device plugin will re-register with kubelet.